### PR TITLE
Add TypeScript documentation, fixes #503

### LINKED ISSUE #503: TypeScript is undocumented
Instead of relying on a separate @types repo, preact ships its own d.ts files, which is very cool. However, Preact does not cover these typings in the official documentation.

This is unfortunate because some things in the typings are nonobvious. For example:

- How do you access props.children in a function component? (answer: either store the function itself in a "let" and type it as a FunctionComponent&lt;props&gt;, or create the function normally and add a field children: ComponentChildren to your props {} type)

- How do you store a JSX element in a variable? (Answer: Import JSX from preact and use JSX.Element)

These things aren't guessable and none of them are exactly the same as the equivalent from @types/React. (I got my answer to the props.children question by finding a blog post explaining FunctionComponent, and then got ComponentChildren from there with IDE detective work. It took a while.)

"Expected behavior": I think Preact should add a page to preactjs.com giving the names it chose to give to the public-facing types in its d.ts and where appropriate (FunctionComponent is sorta nonobvious) how to use them.

### PR DESCRIPTION
**Pull request by ddprrt**
Hi all 👋

This is the a draft addressing #503

The content is mostly adapted from my [TypeScript + React guide](https://fettblog.eu/typescript-react/) I published two years ago, without the fluff and rephrased to suit proper documentation. It can be seen as an accompanying guide to the regular documentation, focusing on how to properly type so the compiler doesn't complain (which can be ... a lot!).

Things done:
- Documenting types for functional and class components
- Proper types for refs and events
- A section on hooks
- Proper typings for context

Things to do:
- A section with exposed types. I currently browse through index.d.ts to get a feeling for all types used and work on scenarios on how to use them in consumer's code.
- Re-check examples if there is a more up to date version. I also want to prevent people and SE crawlers to mistake this as a duplication of my original guide
- If ok, I want to add a section with further reading to Component type design patterns (e.g. https://fettblog.eu/typescript-react-component-patterns/)

Note: For proper syntax highlighting I had to add `tsx` and `typescript` to Prism.js. I used the link in the prism.js comment to get the original configuration, and added two more languages (the new link is again in prism.js)

### CODE REVIEW

**Comment by developit** [comment]
I haven't finished reading through yet, but what I've read so far looks awesome!

Not sure what happened to travis, but ignore the build failure.

File: content/en/guide/v10/typescript.md [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+    if (event.target instanceof HTMLElement) {
+      alert(event.target.tagName); // Alerts BUTTON
+    }
+  }
+
+  render() {
+    return (
+      <>
+        <button onClick={this.handleClick}>{this.props.children}</button>;
+      </>
```
Comment (developit) [inline content/en/guide/v10/typescript.md:202]:
```suggestion
      <button onClick={this.handleClick}>{this.props.children}</button>
```

File: content/en/guide/v10/typescript.md [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+You can restrict event handlers by adding a type annotation for `this` to the function signature as the first argument. This argument will be erased after transpilation.
+
+```tsx
+export class Button extends Component {
+  // Adding the this argument restricts binding
+  handleClick(this: HTMLButtonElement, event: MouseEvent) {
+    event.preventDefault();
+    if (event.target instanceof HTMLElement) {
+      alert(event.target.tagName); // Alerts BUTTON
```
Comment (developit) [inline content/en/guide/v10/typescript.md:194]:
```suggestion
      console.log(event.target.localName); // "button"
```

File: content/en/guide/v10/typescript.md [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+
+  componentDidMount() {
+    // current is of type HTMLAnchorElement
+    console.log(this.ref.current);
+  }
+
+  render() {
+    return <div ref={this.ref}>Foo</div>;
+    //          ~~~
+    //       💥 Error! Ref only can be used for HTMLAnchorELement
```
Comment (developit) [inline content/en/guide/v10/typescript.md:226]:
```suggestion
    //       💥 Error! Ref only can be used for HTMLAnchorElement
```

File: src/config.json [hunk 4]
```diff
@@ -460,6 +459,12 @@
 						"ja": "オプションフック",
 						"de": "Optionen"
 					}
+				},
+				{
+					"path": "/guide/v10/typescript",
+					"name": {
+						"en": "TypeScript"
+					}
```
Comment (developit) [inline src/config.json:467]:
```suggestion
					"name": "TypeScript"
```

**Comment by ddprrt** [comment]
Suggestions committed 🙌

**Review by marvinhagemeister — APPROVED** [review]
Sweet, let's do this! And again thank you so much for the fantastic PR 🙌

**Comment by 38elements** [comment]
@marvinhagemeister 
I think this change includes many repeating, incorrect, unnecessary things.
I think it is incorrect to merge this pull request.

**Pull request merged**